File: PacificMVC/example/src/main/java/com/pacific/example/mvc/controller/MainActivity.java
package com.pacific.example.mvc.controller;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;

import com.pacific.example.R;
import com.pacific.example.mvc.model.MainModel;
import com.pacific.example.mvc.view.MainView;
import com.pacific.mvc.Activity;

public class MainActivity extends Activity<MainModel> {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        model = new MainModel(new MainView(this));
        /** we need to call onCreate() manually*/
        model.onCreate();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main, menu);
        model.setNotify(menu.findItem(R.id.action_notify));
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (model.onOptionsItemSelected(item)) {
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onUserInteraction() {
        super.onUserInteraction();
        model.onUserInteraction();
    }

    @Override
    public void onBackPressed() {
        if (!model.onBackPressed()) {
            super.onBackPressed();
        }
    }

    public boolean hasNotify() {
        return model.hasNotify();
    }
}

File: architecture/example/src/main/java/com/pacific/example/feature/main/MainActivity.kt
package com.pacific.example.feature.main

import android.os.Bundle
import android.support.design.widget.NavigationView
import android.support.design.widget.TabLayout
import android.support.v4.view.GravityCompat
import android.support.v4.view.ViewPager
import android.support.v7.app.ActionBarDrawerToggle
import android.view.Menu
import android.view.MenuItem
import com.pacific.example.R
import com.pacific.example.databinding.ActivityMainBinding
import com.pacific.arch.presentation.activityViewModel
import com.pacific.arch.presentation.contentView
import com.pacific.arch.views.widget.OnTabSelected
import com.pacific.example.base.BaseActivity
import javax.inject.Inject

class MainActivity : BaseActivity(), NavigationView.OnNavigationItemSelectedListener,
        ViewPager.OnPageChangeListener {
    private val binding: ActivityMainBinding by contentView(R.layout.activity_main)
    val model by activityViewModel(MainViewModel::class.java)

    @Inject
    lateinit var mainFragmentAdapter: MainFragmentAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setSupportActionBar(binding.toolbar)
        val toggle = ActionBarDrawerToggle(
                this,
                binding.drawerLayout,
                binding.toolbar,
                R.string.navigation_drawer_open,
                R.string.navigation_drawer_close)
        binding.drawerLayout.addDrawerListener(toggle)
        binding.navView.setNavigationItemSelectedListener(this)
        toggle.syncState()

        binding.viewPager.offscreenPageLimit = mainFragmentAdapter.count
        binding.tabLayout.setupWithViewPager(binding.viewPager)
        binding.tabLayout.addOnTabSelectedListener(object : OnTabSelected() {
            override fun onTabSelected(tab: TabLayout.Tab?) {
                binding.viewPager.setCurrentItem(tab!!.position, false)
            }
        })
        binding.viewPager.clearOnPageChangeListeners()
        binding.viewPager.addOnPageChangeListener(TabLayout.TabLayoutOnPageChangeListener(binding.tabLayout))
        binding.viewPager.addOnPageChangeListener(this)
        binding.viewPager.adapter = mainFragmentAdapter
    }

    override fun onBackPressed() {
        if (binding.drawerLayout.isDrawerOpen(GravityCompat.START)) {
            binding.drawerLayout.closeDrawer(GravityCompat.START)
        } else {
            super.onBackPressed()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        menu.findItem(R.id.action_sync)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.action_search -> {
                onSearchRequested()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    @SuppressWarnings("StatementWithEmptyBody")
    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        val id = item.itemId

        when (id) {
            R.id.nav_camera -> {
            }
            R.id.nav_gallery -> {
            }
            R.id.nav_slideshow -> {
            }
            R.id.nav_manage -> {
            }
            R.id.nav_share -> {
            }
            R.id.nav_send -> {
            }
        }
        binding.drawerLayout.closeDrawer(GravityCompat.START)
        return true
    }

    override fun onPageScrollStateChanged(state: Int) {
    }

    override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    }

    override fun onPageSelected(position: Int) {
    }
}
